### PR TITLE
Button: prevent text wrapping on default variant

### LINKED ISSUE #66049: Button: uniform text wrapping across variants
> It would be interesting to understand why the default button doesn't use `white-space: nowrap` while the primary, secondary, and tertiary variants do. I do understand the `link` variant should allow text to wrap, but why the default does? GIF to illustrate:
> Cc @WordPress/gutenberg-components 
> 
> Image

Originally posted by @afercia in https://github.com/WordPress/gutenberg/issues/65491#issuecomment-2406713485

### Follow-ups

- [ ] Undo the temporary fix in https://github.com/WordPress/gutenberg/pull/66038
- [ ] Look for more places where this change could cause regressions

### PR DESCRIPTION
**Pull request by ciampo**
## What?

Fixes #66049

Prevent text wrapping on the default variant of the `Button` component, in order to align it with all other variants

## Why?

The discrepancy in how text wraps across different variants of the `Button` component leads to inconsistent UI — and it looks like it was an oversight in the initial implementation.

## How?

Tweaked styles

## Testing Instructions

- Make sure that text inside `Button` doesn't wrap to a new line in all variants (including the default variant)
- Smoke test the editor and make sure there are no regressions

## Screenshots or screencast 

| Before (trunk) | After (this PR) |
|---|---|
| Screenshot 2024-10-11 at 16 15 40 |  Screenshot 2024-10-11 at 16 16 12 |

### CODE REVIEW

**Comment by jasmussen** [comment]
Seems fine! 👍 👍

**Review by t-hamano — COMMENTED** [review]
I'm in favor of this change, otherwise the focus outline would cover the text:

image

image

One thing I'm wondering is whether to force `nowrap` on the `link` variation as well.

Only [this variation has auto height](https://github.com/WordPress/gutenberg/blob/ad6e1300a579fd9f8efce4c91f97d4c705aca4c1/packages/components/src/button/style.scss#L249), so the focus outline issue doesn't occur:

| trunk | This PR |
|--------|--------|
| before | after |

**Review by mirka — COMMENTED** [review]
> One thing I'm wondering is whether to force `nowrap` on the `link` variation as well.

Yes. I lean toward keeping the `link` variant untouched.

**Comment by afercia** [comment]
I'd like to suggest to test this change with the `shoeIconLabels` preference set to `true` as many buttons will show text instead of icons. 

One of the first things I noticed is the buttons in the Styles panel header show overlapping text:

Screenshot 2024-10-14 at 08 38 35

This issue is already tracked on https://github.com/WordPress/gutenberg/issues/61761 but making the text `nowrap` makes it even more probleamtic. See also:
https://github.com/WordPress/gutenberg/issues/63251
https://github.com/WordPress/gutenberg/issues/61763
https://github.com/WordPress/gutenberg/pull/61824

Note:
I guess this PR should be rebased to incorporate latest trunk and revert the 'local' fix implemented in https://github.com/WordPress/gutenberg/pull/66038/files ?

**Comment by ciampo** [comment]
> I'd like to suggest to test this change with the `shoeIconLabels` preference set to `true` as many buttons will show text instead of icons.
> 
> One of the first things I noticed is the buttons in the Styles panel header show overlapping text:
> This issue is already tracked on https://github.com/WordPress/gutenberg/issues/61761 but making the text nowrap makes it even more probleamtic. See also:
> - https://github.com/WordPress/gutenberg/issues/63251
> - https://github.com/WordPress/gutenberg/issues/61763
> - https://github.com/WordPress/gutenberg/pull/61824

@afercia thank you for flagging.

It is true that it's an existing situation, but as you say — the changes in this PR would amplify it. We should ideally fix that before merging this change.

More in general, this is exactly the kind of situation that we often warn against when making changes to a component like `Button`. Even a small change can affect the whole app in many ways.

If we wanted to continue with this change, we'd have to check all instances in the editor and make adjustments as needed — and we'd appreciate as much help as possible (at lease in flagging those potential regressions), since we have a limited resources.

> I guess this PR should be rebased to incorporate latest trunk and revert the 'local' fix implemented in [#66038 (files)](https://github.com/WordPress/gutenberg/pull/66038/files) ?

Yup, good point.

**Comment by tyxla** [comment]
I'm a bit concerned that this might break in quite a few places that expect wrapping currently. Example:

|`trunk`|This branch|
|---|---|
|Screenshot 2024-10-14 at 11 49 25|Screenshot 2024-10-14 at 11 49 35

**Comment by ciampo** [comment]
Given the amount of work that it would require to search and fix all those instances, and considering that the fixes may not be trivial, it looks like the approach suggested in this PR is not viable.

A few alternatives:

- we make the change BUT provide a way for consumers to opt out:
  - it could be via a dedicated prop, which has the advantage of abstracting away the internal implementation of the component; 
  - it could be by encouraging applying directly the CSS styles (simpler, but with the potential to break if we ever make changes to the internal DOM structure of the component);

A simpler alternative, inspired by the last point above, is to keep `Button` as-is, and suggest to consumers a workaround to add a custom text wrapping behaviour when needed (eg. to wrap the text inside `Button` inside a `span` with the desired text wrapping behavior).

The last option seems the best compromise in terms of amount of work required (very little), potential regressions (none), and still providing a way to apply the desired text wrapping behaviour. If we ever rework the component and provide a new version (which I think we may at some point), that will be our chance to align text wrapping for all variants.

What do folks think?

**Comment by mirka** [comment]
I personally think the `primary`–`tertiary` variants _shouldn't_ have gotten a `nowrap` style in the first place. I find it a lot more natural for the buttons to have a min-height and wrap by default. If the consumer wants to prevent wrapping, it's usually more natural to control that with `flex`.

So in that sense, I'm not that enthusiastic about adding `nowrap` to the default variant, even if it's for consistency.

**Comment by ciampo** [comment]
@mirka that's a valid point (and I agree with it!)

What do you suggest as the best course of action? Changing text wrapping on those variables? Leaving everything as-is?

**Comment by afercia** [comment]
> I personally think the primary–tertiary variants shouldn't have gotten a nowrap style in the first place. I find it a lot more natural for the buttons to have a min-height and wrap by default.

While I'd agree that buttons should not have a fixed height and should allow text wrapping as native HTML buttons do, I think it's key to make the Gutenberg buttons behave consistently with the Core buttons. [As mentioned on the issue](https://github.com/WordPress/gutenberg/issues/66049#issuecomment-2407071740), the [buttons in core use white-space: nowrap](https://github.com/WordPress/wordpress-develop/blob/655125af1b371eae7c0c0bd154404ab8ed9182c8/src/wp-includes/css/buttons.css#L58) since ages.

**Comment by jameskoster** [comment]
> I find it a lot more natural for the buttons to have a min-height and wrap by default. 

I was going to say the same thing. Not allowing text to wrap can lead to situations where portions of a button label are entirely hidden from vision.

**Comment by afercia** [comment]
To add more context:
Five years ago the Core buttons have been changed to use a `min-height` instead of a fixed height. That was also for accessibility reasons as users can zoom text-only of increase the default operating system font size by a lot, which would make text overlap the boundaries of a button with fixed height. See https://core.trac.wordpress.org/changeset/46243 and previous changes as part of https://core.trac.wordpress.org/ticket/47477

However, that's about the _fixed height_ which is less than ideal and should be avoided. As far as I can tell, the Gutenberg buttons use a _fixed_ height instead (except the is-link variant). When `showIconLabels` is true, some buttons override the height to `auto`. The fixed height is inconsistent with Core and it's an accessibility issue.

Text wrapping is a different matter: Core buttons still use `white-space: nowrap`. I'm not opposed to allow text wrapping but then that should be coordinated with Core and all WordPress buttons should allow wrapping.

However, many parts of the Gutenberg UI use fixed heights. For example, the editor top bar `.editor-header`. If buttons allow text wrapping, they should also have a fluid height otherwise when theyr're in a narrow space this would happen:

Screenshot 2024-10-14 at 15 45 56

On the other side, allowing a fluid height will break the layout of containers with a fixed height e.g. the top bar.

Basically, given the design constraints, either everything should be 'fixed' height or fluid (including cotnainers like the top bar).

**Comment by aaronrobertshaw** [comment]
A lot of great discussion here, thanks everyone 👍 

If I've understood the positions expressed so far, it seems like the general consensus is that we **_should not_** add `nowrap` to the default button as proposed in this PR. Is that a fair assessment?

The introduction of the `nowrap` rule would:
- Make an [existing bug](https://github.com/WordPress/gutenberg/pull/66068#issuecomment-2410348569) worse
- Introduce [new bugs](https://github.com/WordPress/gutenberg/pull/66068#issuecomment-2410488316)
- Generally [go against expectations](https://github.com/WordPress/gutenberg/pull/66068#issuecomment-2410855703) for a button

Keeping the status quo for the default button variant would get my vote if we can provide [viable workarounds](https://github.com/WordPress/gutenberg/pull/66068#issuecomment-2410734216).

The issues of consistency with Core's UI buttons and fixed height/nowrap vs fluid containers etc. are definitely worth discussing. Given the complexity and constraints in charting a path forward there though, perhaps that discussion would be best placed on a dedicated issue for it?

**Comment by afercia** [comment]
> I'm a bit concerned that this might break in quite a few places that expect wrapping currently. Example:

|`trunk`|This branch|
|---|---|
|Screenshot 2024-10-14 at 11 49 25|Screenshot 2024-10-14 at 11 49 35

I share this concern but it's worth mentioning many of these cases are actually ad-hoc implementations that override styles of the default buttons and shouldn't have been implemented this way in the first place. Specifically, the design pattern with "icon on top and text below it' is being discussed on https://github.com/WordPress/gutenberg/issues/62373. The fundamental problem is that the current button variants aren't flexible enough to achieve that design pattern. Either the base button should support many more props and variants (which is hard to justify) or there should be a 'naked' variant of the button that comes with all the functionality but it's completely unstyled.

It's basically a technical debt issue which is costly to solve interms of development time.

**Comment by ciampo** [comment]
Thank you all for the conversation (which is still valid!). I'm just going to close this PR as it looks like the solution proposed is not ideal anyway.
